Feature/tutorial overview page

lovaiible opened the pull request:
### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
This implementation contains an interactive in-app tutorial guide that provides step-by-step explanations on how to use Artemis on the course overview page. 

### Description
Disclaimer: This implementation is the first draft of the guided tour for the course overview page and the foundation for further implementations/optimization for guided tours. The content and layout is subject to change. The guided tour implementation is only visible to users with ADMIN rights and currently won't be started automatically.

When starting the guided tour from the administration menu, the defined DOM elements for the tour step will be highlighted and the tour step window will be displayed on top of an overlay. The user can navigate through the tour, cancel it, and also start it at any other time in any defined language.

### Steps for Testing
1. Log in to ArTEMiS
2. Click on Administration Menu
3. Click on Start Tour
4. Navigate through guided tour

### Screenshots
![guided-tour](https://user-images.githubusercontent.com/33764166/61697914-8739d980-ad38-11e9-8429-d76fbe056fcf.gif)

lovaiible commented on src/main/webapp/app/layouts/navbar/navbar.component.ts:
```diff
@@ -93,4 +102,25 @@ export class NavbarComponent implements OnInit {
+     * that provide tours.
+     */
+    checkGuidedTourAvailability() {
+        if (this.router.url === '/overview') {
```
I currently check if the page is an overview page and then display the 'start overview' menu item in the administration menu. In the future I have to change this since we cannot distinguish all pages only through the router.url

thilo-behnke commented on src/main/webapp/app/overview/overview.component.ts:
```diff
@@ -1,28 +1,50 @@
+            if (guidedTourService.guidedTourSettings && guidedTourService.guidedTourSettings.showCourseOverviewTour) {
+                this.startTour();
+            }
+        }, 500);
```
Shouldn't it be possible in this case to use a timeout of 0?
This should wait for 1 re-render of the template.

If it needs to be 500 to work, it will cause Bugs sooner or later.

I also don't like about it, that the component that uses the guided tour service needs to know about this intricacy of the guided tour (it should be more plug&play).

thilo-behnke commented on src/main/webapp/app/overview/overview.component.ts:
```diff
@@ -57,4 +79,12 @@ export class OverviewComponent {
+    /* Start guided tour for course overview page */
+    public startTour(): void {
+        this.guidedTourService.getOverviewTour().subscribe(tour => {
+            this.overviewTour = tour;
```
Why is this set on the component? Doesn't seem to be used.

thilo-behnke commented on src/main/webapp/app/overview/overview.component.ts:
```diff
@@ -1,28 +1,50 @@
     ) {
         this.loadAndFilterCourses();
+
+        setTimeout(() => {
```
Also I think we should rather do this in the ngOnInit / ngOnChanges and not in the constructor.

thilo-behnke commented on src/main/webapp/app/overview/overview.component.ts:
```diff
@@ -1,28 +1,50 @@
+        }, 500);
+
+        this.subscription = this.guidedTourService.getGuidedTourNotification().subscribe(component => {
+            if (component && component.name === 'overview') {
```
1) This should be a filter in rxjs in a pipe
2) It should rather be possible to subscribe for a certain route and then only get results for this route.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+    private guidedTourNotification = new Subject<any>();
+
+    constructor(public errorHandler: ErrorHandler, private http: HttpClient, private jhiAlertService: JhiAlertService, private accountService: AccountService) {
+        this.getGuidedTourSettings();
```
The constructor should only be used in Angular for initializing variables, everything else should be done in ngOnInit (also for injectable services).

thilo-behnke replied:
This comment is false, the lifecycle hooks are not called for services. I will therefore close this comment.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+
+export type EntityResponseType = HttpResponse<GuidedTourSettings>;
+
+@Injectable()
```
Is it on purpose that this is not injected in root? https://angular.io/guide/dependency-injection#create-an-injectable-service-class

lovaiible replied:
@thilo-behnke I have changed that

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+    public currentTourSteps: TourStep[];
+    public guidedTourSettings: GuidedTourSettings;
+
+    private _guidedTourCurrentStepSubject = new Subject<TourStep>();
```
The underscore pattern is normally only used in languages that don't have privates (python, JS).

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+        return this._currentTour && this._currentTour.steps ? this._currentTour.steps.length : 0;
+    }
+
+    /* Prevents the tour from advancing by clicking the backdrop */
```
This is not proper JSDoc.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+                    // Usually an action is opening something so we need to give it time to render.
+                    setTimeout(() => {
+                        if (this._checkSelectorValidity()) {
+                            this._guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this._currentTourStepIndex));
```
Duplicated code in line 104.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+
+                    // Usually an action is opening something so we need to give it time to render.
+                    setTimeout(() => {
+                        if (this._checkSelectorValidity()) {
```
Duplicated 4 lines below.

lovaiible replied:
@thilo-behnke I have refactored it

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+    /**
+     *  Is last tour step
+     */
+    public get onLastStep(): boolean {
```
I don't like some of the wording that much. E.g. `onLastStep` gives the impression that it is an event and a property. (better `isOnLastStep`)

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+     * @private
+     */
+    private _setFirstAndLast(): void {
+        if (this._currentTour) {
```
This could also just be a getter of the class, the calculation is very easy and more attributes make the class more complicated.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+        document.body.classList.remove('tour-open');
+        this._currentTour = undefined;
+        this._currentTourStepIndex = 0;
+        this._guidedTourCurrentStepSubject.next(undefined);
```
It would be better imo to send null instead of undefined.
Undefined is not checked here by the TS compiler, if the type of the subject is `step | null`, subscribers will know that a null value can happen.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+export class GuidedTourService {
+    public resourceUrl = SERVER_API_URL + 'api/guided-tour-settings';
+
+    public guidedTourCurrentStepStream: Observable<TourStep>;
```
It's never good to give direct access to a class attribute, e.g. making it possible to override it.
I think it would be better to use a getter instead.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+    /**
+     * Send a GET request for the guided tour settings of the current user
+     */
+    private findGuidedTourSettings(): Observable<GuidedTourSettings | undefined> {
```
It would be good to split up the two services and have one service for doing api requests and one for the tour logic.

lovaiible replied:
@thilo-behnke It was more complicated to put this into another service since it then causes circular dependencies. I will keep this in this service for now, since I only need these two calls (GET, PUT). If it gets too big, I will try to move it to another service.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+     */
+    private findGuidedTourSettings(): Observable<GuidedTourSettings | undefined> {
+        return this.http.get<GuidedTourSettings>(this.resourceUrl, { observe: 'response' }).map((res: EntityResponseType) => {
+            if (res.body) {
```
This would mean that if res.body is null or undefined, map returns undefined.
This is not good! Better throw an error or just filter in a pipe.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,367 @@
+    }
+
+    /**
+     * Set orientation of the passed on tour step
```
This JSDoc is very light for such a heavy method.
I wonder in general why there is so much going on this method:
- Why can the orientation be a string?
- What happens if the window doesn't have a certain width?

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/GuidedTourSettingsResource.java:
```diff
@@ -0,0 +1,67 @@
+        log.debug("REST request to update GuidedTourSettings : {}", guidedTourSettings);
+        User currentUser = userService.getUser();
+        currentUser.setGuidedTourSettings(guidedTourSettings);
+        userRepository.save(currentUser);
```
2 things:
1) You are directly using a repository in a resource, you should try to use a service instead
2) I don't like that we have to save the user here, this is very sensitive data and changes to it should be avoided if possible. I would rather have the user_id as foreign-key in the guided_tours.

lovaiible replied:
@thilo-behnke I moved the method to update and save the guided tour settings to the UserService. And for the second point: I discussed it with @krusche that we can we the settings directly in the User table - but we can talk about this if you don't like the solution

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/GuidedTourSettingsResource.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return the guided tour settings
+     */
+    @PutMapping("/guided-tour-settings")
```
As @jpbernius pointed out to me recently, it is a best practice in REST to use `/route/:id` for puts.

lovaiible replied:
@thilo-behnke For me it doesn't seem optimal to add an id here, but I will discuss the REST route with @maxr96 in the future, since he is going to make them consistent.

maxr96 replied:
I agree with @jpbernius that this is the best practice. But in this case only id that we can pass is `userId`, which we always get from the `userService` anyway.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.html:
```diff
@@ -0,0 +1,69 @@
+<div
```
I think it would be good to extract 2-3 components from this component.

lovaiible replied:
@thilo-behnke I removed all orb markup (that we don't use). The component should be oversee-able now, isn't it?

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+    private resizeSubscription: Subscription;
+    private scrollSubscription: Subscription;
+
+    readonly IMAGE = ContentType.IMAGE;
```
Why not just use an enum here? It would be good to have less class attributes.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+    public highlightPadding = 4;
+    public currentTourStep: TourStep | null;
+    public selectedElementRect: DOMRect | null;
+    public isOrbShowing = false;
```
What is an Orb?

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+        this.guidedTourService.guidedTourCurrentStepStream.subscribe((step: TourStep) => {
+            this.currentTourStep = step;
+            if (step) {
+                let hasPermission = true;
```
Better instead:
`const hasPermission = this.accountService.hasAnyAuthorityDirect(step.permission)`

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+     * Modified from https://stackoverflow.com/questions/123999/how-to-tell-if-a-dom-element-is-visible-in-the-current-viewport
+     * @param element
+     */
+    private elementInViewport(element: HTMLElement | null): boolean {
```
This could be a useful util and doesn't have to be specific to this component.

lovaiible replied:
@thilo-behnke That's true. I will implement this in another PR

elkbreeder commented on src/main/webapp/app/layouts/navbar/navbar.component.ts:
```diff
@@ -93,4 +102,25 @@ export class NavbarComponent implements OnInit {
+     * Starts the guided tour of the current component
+     * */
+    startGuidedTour() {
+        if (this.router.url === '/overview') {
```
Call ` checkGuidedTourAvailability()` instead of `this.router.url === '/overview'`

ungaralex commented on src/main/webapp/app/guided-tour/guided-tour.constants.ts:
```diff
@@ -0,0 +1,93 @@
+}
+
+/* Orientation of the tour step position next to the highlighted element */
+export class Orientation {
```
This can be a string enum

lovaiible replied:
@ungaralex I changed that :)

elkbreeder commented on src/main/webapp/i18n/de/global.json:
```diff
@@ -94,7 +94,9 @@
             },
             "language": "Sprache",
-            "noNotifications": "Keine neuen Mitteilungen verfügbar",
+            "guidedTour": "Anleitungstour",
```
Anleitungstour sounds very generic. I think using "Guided Tour" as German Description is just fine. There is also a german wikipedia article about guided tours: https://de.wikipedia.org/wiki/Guided_Tour

elkbreeder commented on src/test/java/de/tum/in/www1/artemis/domain/GuidedTourSettingsTest.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Test
+    public void createFromJson() {
+        String guidedTourSettingsJson = "{\n" + "\t\"showCourseOverviewTour\" : true,\n" + "\t\"showNavigationTour\" : false,\n" + "\t\"showProgrammingExerciseTour\" : true,\n"
```
I would use `Json.createObjectBuilder()` instead: https://docs.oracle.com/javaee/7/api/javax/json/JsonObjectBuilder.html

lovaiible replied:
@elkbreeder I think for this test it's simpler to just use the string

krusche commented:
@lovaiible could you merge develop and work on the feedback? The screencast looks good. 

What I would like to know: Is this tutorial only intended for admins or should it also be available for users who do not see Course Administration and Server Administration?

lovaiible commented:
@krusche Hi Stephan, thank you for the feedback. I am currently working on the feedback and trying to get everything resolved this week. 

Based on the current implementation, the tour is available for every role (USER, TA, ADMIN), but I limited the visibility of the 'Start tour' button in the account menu to ADMIN for now. That limitation should be removed when 'launching' this guided tour feature for production use.

lovaiible commented on src/main/java/de/tum/in/www1/artemis/domain/GuidedTourSettings.java:
```diff
@@ -0,0 +1,103 @@
+    private boolean showNavigationTour = false;
+
+    private boolean showProgrammingExerciseTour = false;
+
```
@elkbreeder I changed that and moved the default initialization to the constructor

lovaiible commented on src/main/java/de/tum/in/www1/artemis/web/rest/GuidedTourSettingsResource.java:
```diff
@@ -0,0 +1,67 @@
+     * @return the guided tour settings
+     */
+    @GetMapping("/guided-tour-settings")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
```
@elkbreeder The tour should be available for all users. I just restricted it on the client-side, therefore I want to keep the authorization here this way :)

lovaiible commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+    /**
+     * Scroll to highlighted element and set tour step
+     */
+    public scrollToAndSetElement(): void {
```
@ungaralex I have refactored it :)

lovaiible commented on src/main/webapp/app/guided-tour/tours/course-overview-tour.ts:
```diff
@@ -0,0 +1,113 @@
+            headlineTranslateKey: 'tour.course-overview.welcome.headline',
+            subHeadlineTranslateKey: 'tour.course-overview.welcome.subHeadline',
+            contentTranslateKey: 'tour.course-overview.welcome.content',
+            imageUrl: 'https://ase.in.tum.de/lehrstuhl_1/images/teaching/interactive/InteractiveLearning.png',
```
@elkbreeder Inserting a relative path is also possible. This content is not fixed yet and will be changed.

lovaiible commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+            return (
+                top >= window.pageYOffset + this.topOfPageAdjustment + scrollAdjustment + this.getStepScreenAdjustment() && top + height <= window.pageYOffset + window.innerHeight
+            );
+        } else {
```
@elkbreeder I have refactored it

lovaiible commented on src/main/java/de/tum/in/www1/artemis/domain/User.java:
```diff
@@ -240,6 +244,17 @@ public void setPersistentTokens(Set<PersistentToken> persistentTokens) {
     }
 
+    public GuidedTourSettings getGuidedTourSettings() {
+        if (this.guidedTourSettings == null || this.guidedTourSettings.isEmpty()) {
```
@elkbreeder I changed that :)

lovaiible commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,498 @@
+     * Subscribe to guidedTourCurrentStepStream and set tour step if the user has the right permission
+     */
+    public ngAfterViewInit(): void {
+        this.guidedTourService.guidedTourCurrentStepStream.subscribe((step: TourStep) => {
```
@elkbreeder I optimized it :)

lovaiible commented:
@thilo-behnke @elkbreeder @ungaralex 

Thank you for your feedback! I tried to incorporate all of them in my latest changes. The major implementations that I did are:
- Moved all guided tour related code from the components to the guided tour service
- Moved <jhi-guided-tour></jhi-guided-tour> to the NavBarComponent, so that we don't have to include it in every component that we want to implement a guided tour for
- Removed the code that handles the orb logic (which we won't need for guided tours in Artemis)
- Added/extended JSDocs
- Added more unit test and 
- Migrate Java tests to JUnit5

maxr96 commented on src/main/java/de/tum/in/www1/artemis/web/rest/GuidedTourSettingsResource.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    /**
+     * UPDATE /guided-tour-settings: update all guided tour settings of the current user
```
Please replace `UPDATE` with `PUT` here

lovaiible replied:
@maxr96 I changed that

maxr96 commented on src/main/java/de/tum/in/www1/artemis/domain/User.java:
```diff
@@ -240,6 +244,17 @@ public void setPersistentTokens(Set<PersistentToken> persistentTokens) {
     }
 
+    public GuidedTourSettings getGuidedTourSettings() {
+        if (this.guidedTourSettings == null) {
```
I would suggest to set `guidedTourSettings` by default for each user. Here inside `@Column` definition you can specify value by setting `columnDefinition` variable. This will work for the new users. For old users you will need to use  `addNotNullConstraint` in Liquibase changelog. You can check how I used it recently for `complaintType` in one of the previous changelogs and more detailed description here https://www.liquibase.org/documentation/changes/add_not_null_constraint.html

lovaiible replied:
@maxr96 thank you for the recommendation. But I think (at least for now) it is better to keep the configuration like this. The settings JSON is an object to change (since this is the first draft) and with the default value I would have to update every single user for every update on the settings.

maxr96 commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,507 @@
+     */
+    @HostListener('document:keydown', ['$event'])
+    handleKeyboardEvent(event: KeyboardEvent) {
+        if (event.code === 'ArrowRight' && this.guidedTourService.currentTourStepDisplay <= this.guidedTourService.currentTourStepCount) {
```
Please use if else or switch here, so that other cases are not checked if first case is true.

lovaiible replied:
@maxr96 I changed that :)

maxr96 commented on src/main/webapp/i18n/en/guided-tour_course-overview.json:
```diff
@@ -0,0 +1,52 @@
+            },
+            "course-admin-menu": {
+                "headline": "Course administration",
+                "content": "Administer and manage your courses and exercises here."
```
You meant administrate here?

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,371 @@
+        this.currentTourSteps = tour.steps;
+
+        // adjust tour steps according to permissions
+        tour.steps.forEach((step, index) => {
```
This could also be a filter operation:

`this.currentTourSteps = tour.steps.filter(step => !step.permission || this.accountService.hasAnyAuthorityDirect(step.permission));`

To me this is a lot clearer to what it means and it does not mutate the existing array.

lovaiible replied:
@thilo-behnke Thanks for the suggestion! I adapted the code as such.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,371 @@
+        this.currentTour = cloneDeep(tour);
+        this.currentTour.steps = this.currentTour.steps.filter(step => !step.skipStep);
+        this.currentTourStepIndex = 0;
+        if (this.currentTour.steps.length > 0 && (!this.currentTour.minimumScreenSize || window.innerWidth >= this.currentTour.minimumScreenSize)) {
```
This if statement either needs a comment or should better be hidden in a function with an explanatory name.

lovaiible replied:
@thilo-behnke I changed that

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,371 @@
+     * @param {step} passed on tour step of a guided tour
+     * @return guided tour step with defined orientation
+     */
+    private setTourOrientation(step: TourStep): TourStep {
```
This is a good example for a method that could be Unit tested:
Input -> Output.

thilo-behnke commented on src/main/webapp/app/layouts/navbar/navbar.component.ts:
```diff
@@ -69,6 +73,14 @@ export class NavbarComponent implements OnInit, OnDestroy {
         }
     }
 
+    checkRouterOnNavigationEnd(): void {
```
Documentation missing, also the naming could be better as it is not implying that it has something to do with the guided tours. Also I have a bad feeling about a subscription that is initialized in the constructor, should be in ngOnInit instead.

lovaiible replied:
@thilo-behnke Thanks for pointing out - I changed that.

thilo-behnke commented on src/main/webapp/app/overview/overview.component.ts:
```diff
@@ -1,15 +1,14 @@
     selector: 'jhi-overview',
     templateUrl: './overview.component.html',
-    styles: [],
+    styleUrls: [],
```
Please just revert this change.

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/domain/GuidedTourSettings.java:
```diff
@@ -0,0 +1,109 @@
+        this.showTextExerciseTour = false;
+    }
+
+    public GuidedTourSettings(boolean showCourseOverviewTour, boolean showNavigationTour, boolean showProgrammingExerciseTour, boolean showQuizExerciseTour,
```
So for every tour we need a new paramter in the constructor?
I think this should be better something that is stored in a Map or so :thinking:

lovaiible replied:
@thilo-behnke I noticed that I don't need this constructor yet (it was implemented in thoughts of possible future use of it). But since there is no need for it now, I will just delete it (and implement it rightfully if it is needed in future).

thilo-behnke commented on src/main/resources/config/liquibase/changelog/20190707025824_changelog.xml:
```diff
@@ -0,0 +1,8 @@
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchan ...
+    <changeSet author="phuong.anh.nguyen (generated)" id="1562461116856-23">
+        <addColumn tableName="jhi_user">
+            <column name="guided_tour_settings" type="varchar(255)"/>
```
I wonder if 255 characters might be not enough at some time... Could be a good choice for longtext.
You might want to check how big it is atm.

lovaiible replied:
@thilo-behnke Yes that's a good point. I changed that to text instead of longtext because that should be more than enough characters.

thilo-behnke commented on src/main/webapp/app/guided-tour/dot-navigation.scss:
```diff
@@ -0,0 +1,62 @@
+.dotstyle {
+    display: inline-block;
+
+    &.dotstyle-scaleup {
```
Wouldn't this be dotstyle-scaleup be `dotstyle--scaleup` in BEM?

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.html:
```diff
@@ -0,0 +1,59 @@
+                    <ul>
+                        <li
+                            *ngFor="let i of guidedTourService.currentTourSteps"
+                            [ngClass]="guidedTourService.currentTourStepDisplay === guidedTourService.currentTourSteps.indexOf(i) + 1 ? 'current' : ''"
```
This should be nicer:
[ngClass.current]="condition"

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,441 @@
+     * @param element: HTMLElement
+     * @return {boolean} if element is in viewport
+     */
+    public elementInViewport(element: HTMLElement | null): boolean {
```
This could also be a good function to Unit test.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -0,0 +1,441 @@
+    /**
+     * @return {number} left position for current tour step
+     */
+    public get leftPosition(): number | null {
```
This could be refactored into a pure function to Unit test it, too.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.constants.ts:
```diff
@@ -0,0 +1,32 @@
+
+/* Orientation of the tour step position next to the highlighted element */
+export enum Orientation {
+    Bottom = 'bottom',
```
Consistency: We usually write enum properties in all caps: `BOTTOM`.

lovaiible commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,371 @@
+     */
+    public skipTour(): void {
+        if (this.currentTour) {
+            if (this.currentTour.skipCallback) {
```
@thilo-behnke The explanation is in the model itself:
`/** Function will be called when tour is skipped */
skipCallback?: (stepSkippedOn: number) => void;`

So basically you can add an optional function that is called as a callback once a tour is skipped.

lovaiible commented:
Hi @thilo-behnke thank you again for the feedback - I implemented all open requests that you had. Regarding the refactoring of mentioned methods for unit testing (since it's an optional request), I would do that if it appears to be necessary for future use cases. Maybe we have to discuss again, if protractor tests wouldn't be more beneficial even though they are prone to changes.

thilo-behnke commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -135,12 +135,11 @@ export class GuidedTourComponent implements AfterViewInit, OnDestroy {
-
         // Set timeout to allow things to render in order to scroll to the correct location
         setTimeout(() => {
+            if (this.isTourOnScreen()) {
```
Nice!

lovaiible commented:
@krusche I made some changes for the database since the last deployment on the test server. Therefore the my initial changelog.xml file should be removed in the db changelog and the guided_tour_settings column should be deleted before this PR is merged and deployed.

krusche commented on src/main/webapp/app/vendor.ts:
```diff
@@ -5,3 +5,252 @@ import 'rxjs/add/operator/map';
 import 'rxjs/add/operator/catch';
 import 'rxjs/add/observable/of';
 import 'rxjs/add/observable/throw';
+import { library } from '@fortawesome/fontawesome-svg-core';
```
I think a merge conflict occurred here. As far as I know, @thilo-behnke moved this into its own module

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/User.java:
```diff
@@ -93,6 +93,10 @@
 
+    @JsonIgnore
+    @Column(name = "guided_tour_settings")
+    private String guidedTourSettings = null;
```
I just thought about this again and I think this would be a bad solution, because of the following reasons:
1) The json string might be quite long and the keys are saved for each user, wasting a lot of unnecessary space
2) The attribute will be loaded every time for each user in each scenario

I think a separate table, that is lazy loaded just in case this information is needed, would be better.
The table would have the userId as foreign key and would have each tutorial as column with a boolean value. Then each user entry could be stored very efficiently in the database, e.g. 1,0,1,0,1 and it would not need around 100 characters of string values per user.

I think this would be modeled with an entity object, e.g. GuidedTourSettings, with booleans. Each user would then have a OneToOne association to such an object and we would use the following annotation for this association.

@OneToOne(fetch = FetchType.LAZY)

Before we deploy this to production, I would ask you to change the entity model. Thanks!

jpbernius replied:
We do not even need a Boolean. If it exist it is true, if it does not exist it is false. The table would map user it and a key which we can be an enum.

jpbernius replied:
Ok, even better: Maybe look at this to store it as a HasMap using JPA: http://www.java2s.com/Tutorials/Java/JPA/0320__JPA_ElementCollection_MapKeyEntity.htm

jpbernius replied:
I keep coming back to this one for some reason. I think as most of the logic here is actually part of the client, I think it would be nice to design the `guidedTourSettings` as a plain key-value store. This way future tours can be implemented on the client-side without any modification to the server side, not even an enum update would be needed. Maybe think of this also when revisiting the decision made here.

lovaiible replied:
@krusche @jpbernius 
Thank you for your feedback.

I can move the guidedTourSettings to a new table and link it to the User table with the user_id as foreign key and only fetch them if necessary.
But I don't know if having an own column for each setting would be good to maintain since we would have to adjust the database for every change. If we keep the settings in a JSON string, however, then we wouldn't have to make server-side adjustments for new guided tours. What do you think about this?

jpbernius replied:
No, the table would be a compound primary key with user if and setting key as keys. And an possible value.
Check the Database die the User to group Mapping.

krusche commented on src/main/webapp/app/guided-tour/guided-tour-step.model.ts:
```diff
@@ -0,0 +1,42 @@
+import { ContentType, LinkType, Orientation, OrientationConfiguration } from 'app/guided-tour/guided-tour.constants';
+
+export interface TourStep {
```
note: for now this is fine. In the future I would like to see a taxonomy here where TourStep is an abstract class and includes several subtypes, e.g. TextTourStep, PictureTourStep, TextTourStep, etc.

krusche replied:
With this we could then avoid most of the `?` in the type definitions

krusche commented on src/main/webapp/app/guided-tour/guided-tour-settings.model.ts:
```diff
@@ -0,0 +1,7 @@
+ * showCourseOverviewTour | showNavigationTour | showProgrammingExerciseTour | showQuizExerciseTour | showModelingExerciseTour | showTextExerciseTour
+ */
+export class GuidedTourSettings {
+    [key: string]: boolean;
```
I have never seen this? Is this an actual map?

jpbernius replied:
Yes, this is a TypeScript Map.

krusche commented on src/main/webapp/i18n/de/guided-tour_course-overview.json:
```diff
@@ -0,0 +1,52 @@
+{
```
If we follow the current naming scheme of json translation files, the filename would be `guidedTourCourseOverview.json`

krusche commented on src/main/webapp/app/guided-tour/guided-tour.module.ts:
```diff
@@ -0,0 +1,32 @@
+    imports: [ArtemisSharedModule],
+    exports: [GuidedTourComponent],
+    entryComponents: [GuidedTourComponent],
+    schemas: [CUSTOM_ELEMENTS_SCHEMA],
```
please remove the CUSTOM_ELEMENTS_SCHEMA.
Have you merged develop? we recently remove all entries of CUSTOM_ELEMENTS_SCHEMA

krusche commented on src/main/webapp/app/guided-tour/guided-tour.service.ts:
```diff
@@ -0,0 +1,396 @@
+            if (this.currentTour.completeCallback) {
+                this.currentTour.completeCallback();
+            }
+            this.updateGuidedTourSettings(this.currentTour.settingsKey, this.currentTourStepDisplay).subscribe(guidedTourSettings => {
```
Does this mean that you save the guided tour settings after each step? I am not sure if this is a good idea, because it could involve a lot of REST calls on the server. Can you save the guided tour setting on the server only if the user closes the guided tour (e.g. by pressing Escape)?

If you want to make sure 100% that intermediate steps are not forgotten, you can always save them in the local browser storage

lovaiible replied:
@krusche The update is only done when the user finishes the guided tour or closing it by clicking the close icon or pressing escape. So that is already implemented :)

krusche replied:
Are you sure? Why do you call the updateGuidedTourSettings(...) then in nextStep()
It is not obvious what you do here. The code is not self-explanatory!

I suggest that you 
1) refactor the code a bit because the method is already 30 : extract a method "finishTour()" that is invoked appropriately and that calls updateGuidedTourSettings(...)
2) Add more comments so that other developers understand the code more easily!

Thanks

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/GuidedTourSettings.java:
```diff
@@ -0,0 +1,50 @@
+ **/
+@Embeddable
+@Table(name = "guided_tour_settings")
+public class GuidedTourSettings implements Serializable {
```
I think this implementation is better. What I still wonder: What happens if a guided tour changes after it was introduced, e.g. when a new feature has been developed and the guided tour needs one additional step in the middle?
